File: Assignment 06/Program 6/prog6.py
#  6. Write a Python program to reverse a string using stack.

class stack:
    def __init__(self,capacity):
        self.stack = []
        self.capacity = capacity
    
    def isEmpty(self):
        if len(self.stack) == 0:
            return 1
        return 0

    def push(self, val):
        if len(self.stack) == self.capacity:
            print("Stack Overflow")
        else:
            self.stack.append(val)
    
    def pop(self):
        if len(self.stack) == 0:
            print("Stack Underflow")
        else:
            return self.stack.pop()
        
def reverse(string):
    s = stack(len(string))
    for i in range(0,len(string)):
        s.push(string[i])
    
    while not s.isEmpty():
        print(s.pop(),end="")

string = input("Enter the string: ")
reverse(string)

File: Data Structure and Algorithm/Assignment 06/Program 6/prog6.c
// 6. Write a C program to reverse a string using stack.

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

typedef struct stack{
    int top;
    int size;
    char *arr;
}stack;

int isFull(stack *ptr){
    if(ptr->top == ptr->size-1)
        return 1;
    return 0;
}

int isEmpty(stack *ptr){
    if(ptr->top == -1)
        return 1;
    return 0;
}

void push(stack *sp, int val){
    if(isFull(sp))
        printf("Stack Overflow");
    else{
        sp->top++;
        sp->arr[sp->top] = val;
    }
}

int pop(stack *sp){
    int item;
    if(isEmpty(sp))
        printf("Stack Underflow");
    else{
        item = sp->arr[sp->top];
        sp->top -- ;
    }
    return item;
}

int peep(stack *sp){
    return sp->arr[sp->top];
}

void reverse_string(char *str){
    stack *sp = (stack*)malloc(sizeof(stack));
    sp->top = -1;
    sp->size = strlen(str);
    sp->arr = (char*)malloc((sp->size+1)*sizeof(char));

    for (int i = 0; str[i] != '\0'; i++)
    {
        push(sp,str[i]);
    }
    while (sp->top != -1)
    {
        printf("%c",pop(sp));
    }
}

int main(){
    char *str;
    printf("Enter the string: ");
    scanf("%s",str);
    reverse_string(str);
    return 0;
}